### PR TITLE
Avoid extra allocation in Encode

### PR DESCRIPTION
**Pull request by yznima**
Avoid extra allocation to increase performance when encoding to QIPC.

Before
---
```
Starting q process on random port
Listening port is  64211
goos: darwin
goarch: amd64
pkg: github.com/sv/kdbgo
BenchmarkEncodeAll-8   	   30000	     48148 ns/op
PASS
Q stdout output: "show `exiting_process;\n"

Q stdout output: `exiting_process

Q stdout output: "exit 0\n"

Q stdout output:
Q stdout error: EOF
Q stderr output:
Q stderr error: EOF
ok  	github.com/sv/kdbgo	2.013s
```

After
---
```
Starting q process on random port
Listening port is  52052
goos: darwin
goarch: amd64
pkg: github.com/sv/kdbgo
BenchmarkEncodeAll-8   	  100000	     20985 ns/op
PASS
Q stdout output: "show `exiting_process;\n"

Q stdout output: `exiting_process
"exit 0\n"

Q stdout output:
Q stdout error: EOF
Q stderr output:
Q stderr error: EOF
ok  	github.com/sv/kdbgo	2.403s
```

### CODE REVIEW

**Comment by yznima** [comment]
@sv Could you review?

File: encode.go [hunk 1]
```diff
@@ -107,18 +107,29 @@ func writeData(dbuf io.Writer, order binary.ByteOrder, data *K) (err error) {
...
-	err = binary.Write(buf, order, dbuf.Bytes())
-	_, err = w.Write(Compress(buf.Bytes()))
+
+	// Then write the qipc encoded data
+	if err := writeData(buf, order, data); err != nil {
+		return err
+	}
+
+	// Now that we have the length of the buffer, create the correct header
+	header = [8]byte{1, byte(msgtype), 0, 0}
```
Comment (sv) [inline encode.go]: no need to allocate second buffer here?

File: encode.go [hunk 2]
```diff
@@ -107,18 +107,29 @@ func writeData(dbuf io.Writer, order binary.ByteOrder, data *K) (err error) {
...
+	order.PutUint32(header[4:], uint32(buf.Len()))
+
+	// Write the correct header to the original buffer
+	b := buf.Bytes()
+	copy(b, header[:])
```
Comment (sv) [inline encode.go]: you can probably write relevant bytes straight to `b`

**Comment by sv** [comment]
Nice PR. Thanks!

File: encode.go (hunk 1 above)
Comment (yznima) [reply encode.go]: done

File: encode.go (hunk 2 above)
Comment (yznima) [reply encode.go]: @sv I tried it but it didn't improve performance. If you are ok, I think for readability, we should keep the header array

**Comment by yznima** [comment]
@sv Could you re-review

**Comment by sv** [comment]
Could you update description with new test timing if it has changed?

**Comment by yznima** [comment]
> Could you update description with new test timing if it has changed?

Done

**Comment by yznima** [comment]
> Could you update description with new test timing if it has changed?

@sv Done

**Comment by yznima** [comment]
@sv ready for merge

**Pull request merged**